### PR TITLE
Create a Spec for v0.3 that can be used to implement PAST in other languages

### PR DESCRIPTION
**Pull request by paragonie-scott**
This is part of the work needed for #16

### CODE REVIEW

**Comment by paragonie-scott** [comment]
Okay, this is pencils-down for me for the time being. Once this is merged, I'll get `v0.3.0` tagged and released, and we can focus on the formal RFC.

File: docs/01-Protocol-Versions/Version1.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+   * Empty: return "`h` || base64url(`n` || `c` || `t`)"
+   * Non-empty: return "`h` || base64url(`n` || `c` || `t`) || `.` || base64url(`f`)"
+   * ...where || means "concatenate"
+
+## Decrypt
+
+Given a message `m`, key `k`, and optional footer `f`
+(which defaults to empty string):
+
+1. If `f` is not empty, verify that the value appended to the token matches `f`,
```
Comment (ericchiang) [inline docs/01-Protocol-Versions/Version1.md]: If the footer is supposed to be known ahead of time, why include it in the token instead of just using it in the MAC calculation?

File: docs/01-Protocol-Versions/Version1.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+footer `f` (which defaults to empty string):
+
+1. If `f` is not empty, verify that the value appended to the token matches `f`,
+   using a constant-time string compare function. If it does not, throw an exception. 
+2. Verify that the message begins with `v1.public.`, otherwise throw an exception.
+   This constant will be referred to as `h`.
+3. Decode the payload (`sm` sans `h`, `f`, and the optional trailing period
+   between `m` and `f`) from base64url to raw binary. Set:
+   * `s` to the rightmost 256 bytes
+   * `m` to the leftmost remainder of the payload, excluding `t`
```
Comment (ericchiang) [inline docs/01-Protocol-Versions/Version1.md]: Is "excluding `t`" wrong? I don't see `t` defined in the signing section.

File: docs/01-Protocol-Versions/Version2.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+   ```
+   c = crypto_aead_xchacha20poly1305_encrypt(
+       message = m
+       aad = preAuth
+       nonce = n
+       key = k
+   );
+   ```
+6. If `f` is:
+   * Empty: return "`h` || base64url(`c`)"
```
Comment (ericchiang) [inline docs/01-Protocol-Versions/Version2.md]: Need to include the nonce here?

```
`h` || base64url(`n` || `c`)
```

File: docs/README.md [hunk 4]
```diff
@@ -37,6 +37,52 @@ See [Protocol Versions](01-Protocol-Versions) for specifics.
...
+Developers who are already familiar with JSON Web Tokens (JWT) should be able
+to, intuitively, use PAST in their software with minimal friction.
+
+Additionally, developers who are not already familiar with JWT should be able
+to pick up PAST and use it successfully without introducing security flaws
+into their application.
+
+### 3. Flexibility
+
+Although our library aims to maximize JWT compatibility, out cryptography
```
Comment (ericchiang) [inline docs/README.md]: JWT have an equivalent to this, a JWS. To say that PAST can handle more payload types than JOSE seems disingenuous.

File: docs/01-Protocol-Versions/Version1.md (hunk 1 above)
Comment (carnage) [reply docs/01-Protocol-Versions/Version1.md]: I was going to query this as well; one of the initial use cases for a footer was the possibility to append a "key-id" or similar (which wouldn't be encrypted in local tokens) in which case it would have been processed by the app prior to passing to the library to grab the correct key.

If someone has modified it, the mac should fail

File: docs/01-Protocol-Versions/Version1.md (hunk 2 above)
Comment (carnage) [reply docs/01-Protocol-Versions/Version1.md]: t is the signature I believe

File: docs/01-Protocol-Versions/Version1.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
+# PAST Version 1
+
+## GetNonce
+
+Given a message (`m`) and a nonce (`n`):
+
+1. Calculate HMAC-SHA384 of the message `m` with `n` as the key.
```
[the remark this answers is not in the document]
Comment (paragonie-scott) [reply docs/01-Protocol-Versions/Version1.md]: Nonce-misuse resistance. If an implementation has an insecure RNG, this adds a layer of insurance against nonce-misuse, which breaks CTR mode.

File: docs/01-Protocol-Versions/Version2.md (hunk 3 above)
Comment (paragonie-scott) [reply docs/01-Protocol-Versions/Version2.md]: Yep!

**Comment by carnage** [comment]
I was also looking over the common.md part of the docs; wondering if a 64bit length is a bit overkill. The intended use case for this is an authentication token containing information about identity, validity (eg who issued it, who can use it, time period it's valid for) and maybe permissions/roles, even encoded as XML that should be a pretty small data structure. 32 bits for length should be easily sufficient (4GB tokens YAY!) 16 bits would also probably work unless a theoretical post quantum v3 requires longer signatures

File: README.md [hunk 6]
```diff
@@ -21,7 +21,10 @@ for secure stateless tokens.
 Unlike JSON Web Tokens (JWT), which gives developers more than enough rope with which to
 hang themselves, PAST only allows secure operations. JWT gives you "algorithm agility",
 PAST gives you "versioned protocols". It's incredibly unlikely that you'll be able to
-use PAST in [an insecure way](https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries). 
+use PAST in [an insecure way](https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries).
+
+> **Caution:** Neither JWT nor PAST were designed for
+> [stateless session management](http://cryto.net/~joepie91/blog/2016/06/13/stop-using-jwt-for-sessions/).
```
Comment (carnage) [inline README.md]: Perhaps copy some of the points over and reference the blog for further reading

File: docs/01-Protocol-Versions/Version1.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (ericchiang) [reply docs/01-Protocol-Versions/Version1.md]: If an implementation has an insecure RNG there are way bigger issues, right?

File: docs/README.md [hunk 7]
```diff
@@ -37,6 +37,52 @@ See [Protocol Versions](01-Protocol-Versions) for specifics.
...
+protocols operate over binary strings rather than JSON objects, thereby
+allowing PAST to be used with other serialization standards.
+
+To define these other wire formats, the name of the encoding should be
+suffixed:
+
+* **PAST-JSON** is the default format for payload serialization
+* **PAST-Protobuf** means PAST cryptography using Protobuf serialization
+* **PAST-Raw** uses the cryptography features directly over raw strings
+* **PAST-XML** means PAST cryptography using XML serialization
```
Comment (carnage) [inline docs/README.md]: I think it may be better to drop this section and stick with a single packed data format (JSON) if PAST is meant to be a crypto hardened replacement for JWT, it would make sense to avoid scope creep and stick as closely as possible to the capabilities of JWT

**Comment by paragonie-scott** [comment]
I specified 64-bit to prevent [weird bugs like this](https://threatpost.com/bugs-in-signal-messaging-app-corrupt-attachments-crash-app/120643/).

File: docs/README.md (hunk 7 above)
Comment (paragonie-scott) [reply docs/README.md]: OK.

File: README.md (hunk 6 above)
Comment (paragonie-scott) [reply README.md]: I really don't want to spend more bytes than I have to on why that's a bad idea.

File: docs/01-Protocol-Versions/Version1.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (paragonie-scott) [reply docs/01-Protocol-Versions/Version1.md]: If your keys were generated securely outside the context of the system being used, but your RNG is still busted, then at least we can ensure NMR for our protocol.

**Comment by carnage** [comment]
Will re-read that in the morning; but I don't think that attack vector would be valid here; we could just reject tokens which are greater than X in size. If someone is sending a 4GB authentication token to your server, you can be fairly certain something weird is going on...

**Comment by paragonie-scott** [comment]
Keep in mind: This never goes into the token. The PAE format is just used for concatenating data before passing to HMAC/Ed25519. It won't increase the token size, etc.

**Comment by paragonie-scott** [comment]
Okay, I'm satisfied with the current state of affairs. Thanks @carnage and @ericchiang for reviewing the documentation. 👍
